plugins/@grouparoo/klaviyo/src/lib/export/exportRecord.ts

```diff
@@ -0,0 +1,159 @@
+}
+
+export const handlePersonChanges: ExportRecordPluginMethod = async ({
+  appOptions,
+  syncOperations,
+  export: {
+    newRecordProperties,
+    oldRecordProperties,
+    newGroups,
+    oldGroups,
+    toDelete,
+  },
+}) => {
+  if (Object.keys(newRecordProperties).length === 0) {
+    return { success: true };
+  }
+
+  const client = await connect(appOptions);
+
+  const newEmail = newRecordProperties["email"];
+  const oldEmail = oldRecordProperties["email"];
+
+  if (!newEmail) {
+    throw new Error(`newRecordProperties[email] is a required mapping`);
+  }
+
+  const newFoundId = await client.findProfileIdByEmail(newEmail);
+
+  let oldFoundId = null;
+  if (oldEmail && oldEmail !== newEmail) {
+    oldFoundId = await client.findProfileIdByEmail(oldEmail);
+  }
+
+  const foundId = newFoundId || oldFoundId;
+  if (toDelete) {
+    if (!syncOperations.delete) {
+      throw new Errors.InfoError(
+        "Destination sync mode does not delete records."
+      );
+    }
+    if (foundId) {
+      await client.dataPrivacy.requestProfileDeletion({
+        identifier: foundId,
+        id_type: "person_id",
+      });
+    }
+    return { success: true };
+  }
+  const payload = await makePayload(oldRecordProperties, newRecordProperties);
+  if (foundId) {
+    // Update existing Person
+    if (!syncOperations.update) {
+      throw new Errors.InfoError(
+        "Destination sync mode does not update existing records."
+      );
+    }
+    await client.profiles.updateProfile({
+      profileId: foundId,
+      properties: payload,
+    });
+  } else {
+    if (!syncOperations.create) {
+      throw new Errors.InfoError(
+        "Destination sync mode does not create new records."
+      );
+    }
+
+    // Create new Person
+    await client.public.identify({
+      email: newEmail,
+      properties: payload,
+      post: true,
+    });
+  }
+  await handleGroupsChanges(client, newEmail, oldGroups, newGroups);
+  return { success: true };
+};
+
+export const exportRecord = async (args) => {
+  try {
+    return await handlePersonChanges(args);
+  } catch (error) {
+    // look for the rate limit exceeded status code.
+    if (error?.response?.status === 429) {
+      const retryIn = Math.floor(Math.random() * 10) + 1;
+      return { error, success: false, retryDelay: 1000 * retryIn };
+    }
+    throw error;
+  }
+};
+
+export async function makePayload(oldRecordProperties, newRecordProperties) {
+  const payload = {};
+
+  // set record properties, including old ones
+  const newFields = Object.keys(newRecordProperties);
+  const oldFields = Object.keys(oldRecordProperties);
+  const allFields = new Set([...newFields, ...oldFields]);
+
+  // get fields
+  const fieldKeys = await getKnownFieldMap();
+  for (const fieldName of allFields) {
+    const value = newRecordProperties[fieldName];
+    const klaviyoKey = fieldKeys[fieldName];
+
+    payload[klaviyoKey ? klaviyoKey : fieldName] =
+      formatVar(value) ?? "[deleted]";
```

Comment:
maybe set to null or empty string if that works.

Reply:
It doesn't work, any `falsy` value will be ignored!
for them, null or empty string means keep it untouched!!
Basically, they don't have such "clear field" functionality. and that was the best solution I've thought about to deal with that weird behavior.

Reply:
I bet it will fail on date and integers, though. 🤔

Reply:
No dates or integers, everything is string here. they don't have different data types! Only strings.
and `formatVar` returns `null` only when the field is set `undefined` which means "erase the value of it".

Reply:
I'm pretty sure `formatVar` is never going to return `undefined` here, so this will never be `[deleted]`.
I think you want something like:
```
if (value===undefined) value = "[deleted]";
payload[fieldName] = formatVar(value);
```

or otherwise handling undefined before formatting.

``

Reply:
Yup, it will never return `undefined`, it returns `null` instead if the value isn't set in the new data.
That's why I'm using `?? (Nullish Coalescing)` to convert `null` to `"[deleted]"`

Reply:
I think you want to have `"deleted"` if it was truly deleted though. If someone's last name (or whatever setting) is `null` in grouparoo, it's probably not right that it ends up in Klaviyo as `"deleted"`.

Reply:
Aha, I got the point

Reply:
Let's try this `$unset` approach and see if it works. https://community.klaviyo.com/apis-40/set-custom-property-to-null-blank-or-remove-it-1501

Reply:
Random thought: I guess we should also use `$unset` if it's falsy at all inlcuding empty string if that's how Klaviyo works. so maybe `formatVar` should trim and then return null if it's empty.